apps/prairielearn/src/ee/lib/billing/plan-grants.ts

```diff
@@ -0,0 +1,58 @@
+import { type Response } from 'express';
+
+import { CourseInstanceSchema, InstitutionSchema, UserSchema } from '../../../lib/db-types';
+import { getEnrollmentForUserInCourseInstance } from '../../../models/enrollment';
+import {
+  getPlanGrantsForContextRecursive,
+  getPlanNamesFromPlanGrants,
+  getRequiredPlansForCourseInstance,
+} from '../../lib/billing/plans';
+import { planGrantsMatchPlanFeatures } from './plans-types';
+
+export async function checkPlanGrants(res: Response) {
+  // We'll only check plan grants for course instances, as students can't
+  // currently ever access a course directly. And even if they could, plan
+  // grans aren't associated with courses.
+  if (!res.locals.course_instance) {
+    return true;
+  }
+
+  // We won't check plan grants if the user has a specific role in the course
+  // or course instance. We always grant instructor-like users access to all
+  // features.
+  if (
+    res.locals.authz_data.course_role !== 'None' ||
+    res.locals.authz_data.course_instance_role !== 'None'
+  ) {
+    return true;
+  }
+
+  const institution = InstitutionSchema.parse(res.locals.institution);
+  const course_instance = CourseInstanceSchema.parse(res.locals.course_instance);
+  const user = UserSchema.parse(res.locals.user);
+
+  const enrollment = await getEnrollmentForUserInCourseInstance({
+    // Note that this takes into account user overrides set by instructors.
+    // This means that instructors impersonating a student will see the same
+    // behavior that students themselves would see. If the instructor wants to
+    // bypass any plan grant checks, they can use the "Student view without
+    // access restrictions" option.
```

Comment:
Won't instructors already have been short-circuited by the above checks for `course_role`? I can't see how they get to this code.

Reply:
If I'm understanding your concern correctly: remember that `authz_data` contains the *effective* information. Even if a user does *actually* have instructor permissions, they may have requested to emulate *not* having those permissions. In that case, `course_role` and `course_instance_role` would in fact be none, so we'd fall through to here.

Reply:
Ah cool, ok, makes sense to me now!

Reply:
Yay! I'll add an inline comment to remind us of this fact.

Reply:
Comment added in c1994b5.